Async DNS search domain support

jroper opened the pull request:
Fixes #25825

Provides search domain support for Async DNS.

There are two parts here, first is the search domain support itself, which is relatively straight forward - on lookup, create a list of hosts to attempt to resolve based on the search domains configuration (including the host with no postfix), exclude all those hosts that have less dots than the configured ndots, and then search until you find one that returns any records. In this case the search domains are applied for each nameserver attempt, so if a particular nameserver times out, it won't attempt to search any more domains on that nameserver.

The second part is in actually detecting the system configuration. I've only implemented partial support here, it reads /etc/resolv.conf, and nothing else. There are ways to detect it on other platforms, there are also environment variables that are supposed to be honored on Linux and other *nix's, but to keep this change manageable, I've focussed on just supporting the minimum to meet the Kubernetes service discovery use case, and documented the limitations.

jroper committed:
Async DNS search domain support

Fixes #25825

Provides search domain support for Async DNS.

There are two parts here, first is the search domain support itself,
which is relatively straight forward - on lookup, create a list of hosts
to attempt to resolve based on the search domains configuration
(including the host with no postfix), exclude all those hosts that have
less dots than the configured ndots, and then search until you find one
that returns any records. In this case the search domains are applied for
each nameserver attempt, so if a particular nameserver times out, it
won't attempt to search any more domains on that nameserver.

The second part is in actually detecting the system configuration. I've
only implemented partial support here, it reads /etc/resolv.conf, and
nothing else. There are ways to detect it on other platforms, there are
also environment variables that are supposed to be honored on Linux and
other *nix's, but to keep this change manageable, I've focussed on just
supporting the minimum to meet the Kubernetes service discovery use case,
and documented the limitations.

raboof commented:
PLS BUILD

dwijnand commented on akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala:
```diff
@@ -44,6 +46,50 @@ private[dns] final class DnsSettings(system: ExtendedActorSystem, c: Config) {
 
+  private lazy val resolvConf: Option[ResolvConf] = {
+    val etcResolvConf = new File("/etc/resolv.conf")
+    // Avoid doing the check on Windows, no point
```
Any point in opting out if this is conditional on `/etc/resolv.conf` existing?

jroper replied:
What if it did exist on Windows? Would using it be a good idea? That's actually the real reason, if it did exist, reading it would probably not be the right thing to do.

dwijnand replied:
I would expect nothing to exist under `/` on Windows.  And if it did exist (for whatever reason), I'd expect it using it would be the right thing.

dwijnand commented on akka-docs/src/main/paradox/io-dns.md:
```diff
@@ -5,9 +5,13 @@
 
+Additionally, while search domains are supported through configuration, detection of the system configured
+[Search domains](https://github.com/akka/akka/issues/25825) is only supported on systems that provide this 
+configuration through a `/etc/resolv.conf` file, i.e. it isn't supported on Windows or OSX, and none of the 
```
There's an `/etc/resolv.conf` on my Mac, though it comes with the header:

```
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
```

(Also s/OSX/macOS/)

jroper replied:
Yeah, I was aware that that existed. The Netty code that does this implies that Darwin actually has files in `/etc/resolver/*` that can be read, but the documentation it links to 404s.

dwijnand replied:
I don't have a `/etc/resolver/` directory at all.  🤷‍♂️

raboof commented:
Sorry about the jenkins noise, we were just updating the slaves.

It seems some tests are now failing because they are missing the 'search-domains' configuration option.

ignasi35 commented on akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala:
```diff
@@ -44,6 +46,50 @@ private[dns] final class DnsSettings(system: ExtendedActorSystem, c: Config) {
+        case Success(value) ⇒ Some(value)
+        case Failure(exception) ⇒
+          if (system.log.isWarningEnabled) {
+            system.log.error(exception, "Error parsing /etc/resolv.conf, ignoringto")
```
check `isWarningEnabled` but `log.error` is used. Is that on purpose?

ignasi35 replied:
Also, the message seems incomplete: `"..., ignoringto"`

jroper replied:
It's on purpose - the only way to log an exception in Akka (and get that exception to be passed as an exception to the underlying log framework and logged according to that log frameworks configuration, eg with the number of stack trace lines configured by that framework) is to log it at `error`, since Akka's logger doesn't have `warn`, `info` or `debug` methods that accepts exceptions. So the next best thing is to check that the log level that you actually want it logged at is enabled before logging it at error - this is done throughout Akka.

And the message is a mistake, I'll fix.

dwijnand replied:
> So the next best thing is to check that the log level that you actually want it logged at is enabled before logging it at error - this is done throughout Akka.

How confusing.  I wonder why this is the case.

chbatey replied:
Not adding methods to a public interface. Typed has stack traces for levels other than error

ignasi35 commented on akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala:
```diff
@@ -44,6 +46,50 @@ private[dns] final class DnsSettings(system: ExtendedActorSystem, c: Config) {
+        c.getStringList("search-domains").asScala.toList
+      case _ ⇒ throw new IllegalArgumentException("Invalid type for search-domains. Must be a string or string list.")
+    }
+  }
```
Is it an Akka-specific requirement to support both `STRING` and `LIST` config types? Could we _just_ make the default `[]` and require this setting to always be a `List[String]`?

jroper replied:
I've basically just copied the code above for handling `nameservers` config here, so it's consistent with the `nameservers` config which I think makes sense. There's actually a good reason for supporting both, we have to still read it and check the type, since that's required to read a value of `default`. Since we're reading a single string value anyway, may as well use that value if it's not `default`.

ignasi35 commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -1061,6 +1061,17 @@ akka {
+        # Configures the list of search domains.
+        # Defaults to a system dependent lookup (on Unix like OSes, will attempt to parse /etc/resolv.conf, on
+        # other platforms, will not make any attempt to lookup the search domains). Set to a single domain, or
+        # a list of domains, eg, [ "example.com", "example.net" ].
```
👍 I like the example values for expected syntax (e.g. no trailing `.`: `example.net.`)

ignasi35 commented on akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala:
```diff
@@ -44,6 +46,50 @@ private[dns] final class DnsSettings(system: ExtendedActorSystem, c: Config) {
+          case _         ⇒ throw new IllegalArgumentException("Invalid value for ndots. Must be the string 'default' or an integer.")
+        }
+      case ConfigValueType.NUMBER ⇒
+        c.getInt("ndots")
```
Should also check it's a positive number.

Going further, does `0` make sense? (to enforce the lookup?) If so, we should document the meaning of `0` on `reference.conf`.

jroper replied:
Zero makes sense (eg, see https://github.com/netty/netty/issues/5570).

ignasi35 commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -1061,6 +1061,17 @@ akka {
+        # Configures the number of dots that should be in a domain before a lookup will be done.
+        # Defaults to a system dependent lookup (on Unix like OSes, will attempt to parse /etc/resolv.conf, on
+        # other platforms, will default to 1).
+        ndots = default
```
TIL `ndots` comes from the `/etc/resolv.conf`. It'd be great to mention that (with a link to https://linux.die.net/man/5/resolver).

My first impression when I saw this setting was: "Hmm, I think this name is wrong, maybe `num-of-dots` or s/thing?" but it makes sense to me now.

ignasi35 reviewed (commented):
Few comments here and there.

jroper commented:
There's a problem that I've just found with this PR - resolving localhost is now broken if any search domains are defined and ndots is greater than 0. Technically, the async dns resolver has never properly supported resolving localhost - you're not supposed to query a DNS server for localhost, it's supposed to always be present in the `/etc/hosts` file. However, in practice, due to recommendations from [RFC1912](https://www.ietf.org/rfc/rfc1912.txt), nameservers usually do resolve localhost to`127.0.0.1`, to "to help eliminate accidental queries for broadcast or local address to be sent off to the root nameservers". Google's nameservers don't resolve localhost though.

So, what this means, is people that upgrade to this that are relying on the async-dns resolver to resolve `localhost` (probably lots of people in development) who happen to be using a nameserver that resolves it are going to find that this breaks if they have a `domain` or `search` element in their `resolv.conf` file. Working around is easy enough, set `search-domains = []` or `ndots = 0` in your `application.conf`, the former effectively disables the new code paths added by this PR. But we should potentially consider either adding special handling for `localhost`, or supporting `/etc/hosts` properly.

jroper commented:
I'm happy to implement the support parsing and supporting `/etc/hosts` if you think that's necessary before this PR is merged, I've taken a look at Netty's code and it looks straight forward enough to parse - assuming the parsing approach I'm taking (completely ad-hoc using things like `String.split` and `RichString.span`, not really taking performance or memory usage into consideration), is ok. I considered using Scala parser combinators or something similar but that would mean adding a new dependency to akka-actor which in my opinion is way too overkill for parsing these simple files at startup.

Once it's parsed, I think the idea is that it gets treated like its own nameserver, one that sits first in the list. So the simplest way to implement this is to create a new actor that implements the same protocol as the current DnsClient actors but is backed by a map.

Also something to consider, both with `/etc/hosts`, and `/etc/resolv.conf`, is whether we want to be able to pick up changes to these files. Given that our current method of consuming the system nameservers doesn't pick up changes (if I'm not mistaken I think the JVM doesn't monitor these files for changes itself, so nothing short of restarting the process will get what we need), it's probably not necessary to support it (it would be weird if we picked up changes to the search domains in `/etc/resolv.conf` but not the nameservers), but it's something to consider. Now that we've got code to parse `/etc/resolv.conf`, we can very feasibly implement a polling task to pick up changes.

jroper commented:
So - should we go with implementing `/etc/hosts` support too, or leave resolving `localhost` in a broken (as it technically always has been) but different broken state?

chbatey commented on akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala:
```diff
@@ -44,6 +46,50 @@ private[dns] final class DnsSettings(system: ExtendedActorSystem, c: Config) {
+  private lazy val resolvConf: Option[ResolvConf] = {
+    val etcResolvConf = new File("/etc/resolv.conf")
+    // Avoid doing the check on Windows, no point
+    if (sys.props.get("os.name").exists(os ⇒ os.startsWith("Windows") || os.startsWith("WINDOWS"))) {
```
`Helpers.isWindows`

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/ResolvConfParser.scala:
```diff
@@ -0,0 +1,66 @@
+   */
+  def parseFile(file: File): Try[ResolvConf] = {
+    Try {
+      parseLines(Files.lines(file.toPath).iterator().asScala)
```
Blocking io on the default dispatcher

chbatey replied:
I guess this is okay as it is done once and on initialisation of DNS extension

jroper replied:
That was my thinking, it's in the same class of thing as reading `application.conf` on startup.

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/ResolvConfParser.scala:
```diff
@@ -0,0 +1,66 @@
+  def parseLines(lines: Iterator[String]): ResolvConf = {
+    // A few notes - according to the spec, search and domain are mutually exclusive, the domain is used as the
+    // sole search domain if specified. Also, if multiple of either are specified, then last one wins, so as we
+    // parse the file, as we encounter either a domain or search element, we replace this list with what we find.
```
:+1:

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala:
```diff
@@ -51,10 +51,10 @@ private[io] final class AsyncDnsResolver(
 
   override def receive: Receive = {
```
Let's add the search domains to the debug log at the top of this file

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala:
```diff
@@ -86,6 +86,33 @@ private[io] final class AsyncDnsResolver(
 
+  private def resolveWithSearch(name: String, requestType: RequestType, resolver: ActorRef): Future[DnsProtocol.Resolved] = {
+    if (settings.SearchDomains.nonEmpty) {
+      val nameWithSearch = settings.SearchDomains.map(sd ⇒ name + "." + sd)
```
what is one of these doesn't have enough dots?

jroper replied:
Originally I was just using a filter, but after rereading the documentation, it makes it clear that this only applies to the name being searched, see [here](http://man7.org/linux/man-pages/man5/resolv.conf.5.html):

> Sets a threshold for the number of dots which must
> appear in a name given to res_query(3) (see
> resolver(3)) before an initial absolute query will be
> made.  The default for n is 1, meaning that if there
> are any dots in a name, the name will be tried first as
> an absolute name before any search list elements are
> appended to it.  The value for this option is silently
> capped to 15.

The "initial" I think is the key there, ndots only affects the initial query, not subsequent queries based on search domains. Furthermore, the logic here is the same as the logic used by netty.

Now, I've just done a test on my linux box, and it's hard to work out exactly what it's doing, but for example if I set:

```
search com
options ndots:2
```

Then `host github` gives me `github.com`, so `ndots:2` isn't preventing `github.com` to be searched. But, if I search `host github.com`, then I get `github.com.com`, (which redirects to `com.com` and is a site setup to capture accidental double coms in domain names and try and monetise that traffic). Interestingly, if I search `host akka.io`, I get `akka.io`. So presumably, `akka.io.com` was searched (it doesn't resolve), so it then tried `akka.io`. So it appears that the implementation in Linux is actually not quite as the documentation states, rather, ndots controls whether the hostname passed in is searched first or last. We could use that behaviour too, and I think that would have the effect of fixing (or at least, ensuring that the old behaviour is still implemented) the `localhost` issue.

jroper replied:
I've modified the code now to put the host at the end of the search list if it fails the ndots test, and I also added a test to ensure resolving `localhost` works, which passes because the nameserver we use in the integration test resolves `localhost`. See https://github.com/akka/akka/compare/a02a6375d638eee1864cdf0a0c080d57c16f261f..e75212eed78b48a82457d66252c6861c32ad577d for changes.

chbatey replied:
:+1:

chbatey reviewed (commented):
Great stuff, I've tested this with Kubernetes Bootstrap using DNS without configuring including any of the namespace / cluster name in the configuration and it works.

I've added some follow up issues which you can link to in the docs if you want to:
* Windows hack for search list: https://github.com/akka/akka/issues/26002
* LOCALDOMAIN env var https://github.com/akka/akka/issues/26001

jroper commented:
Putting in the cache without a TTL may work, except for one thing (and this also applies to making it just another nameserver), the problem is that if ndots is 1 (the default), and `localhost` is looked up, the current code will not attempt to search the cache for `localhost`. So I think what we need to do is make the ndots check turn searching the nameserver on/off, but always search the cache regardless.

jroper commented:
Actually, I've just realised another problem with this PR, that is, since NXDOMAINS aren't cached, any queries that aren't resolved from the first element in the search list will always have to go through the search list until they reach the domain that is cached. I'll fix this.

jroper commented:
I think I've found a bug, if a lookup is done for both IPv4 and IPv6 (which is the default), but only one of those resolves (which would be quite common, eg, an IPv4 only network), then only that lookup is cached, the other isn't. So, although you might be getting the cached IPv4 result, you still have to wait for the IPv6 lookup with no records on each request. What I'll do is modify the cache to cache IPv4 and IPv6 lookups together, so it'll cache by `RequestType` (ie, `Srv` or `Ip(ipv4: Boolean, ipv6: Boolean)`), rather than by the individual queries/answers.

jroper committed:
Fixed Async DNS caching

Ensure that queries for both IPv4 and IPv6 are cached together so that
subsequent requests don't have to search ones that return nothing. Also
fixed caching for search domains so that the caching is done against the
original name.

jroper commented:
So I just pushed something that changes the caching to fix the above. Also, the cache check is done first, before iterating through any resolvers or anything, this means that caching is done against the original name looked up, rather than against the search domain. So if the search domain is `example.org`, and you do a look up of `foo`, and then of `foo.example.org`, these will each have their own cache entry, which is probably the right thing anyway.

chbatey commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala:
```diff
@@ -39,7 +37,7 @@ private[io] final class AsyncDnsResolver(
   val nameServers = settings.NameServers
 
-  log.debug("Using name servers [{}]", nameServers)
+  log.debug("Using name servers [{}] and search domains [{}] with ndots={}", nameServers, settings.SearchDomains, settings.NDots)
```
:+1:

chbatey commented:
This is looking good. I'd be keen to get the cache changes in before doing https://github.com/akka/akka/issues/25948

What do you think is left @jroper ? Looks ready to merge

jroper commented:
I think it's ready. With regards to the mima check failing, it's all internal API, but I think I could put all the missing classes/methods back, with them deprecated (or throwing). Should I do that, or add problem filters?

raboof commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala:
```diff
@@ -70,10 +80,10 @@ private[io] final class AsyncDnsResolver(
           case NonFatal(t) ⇒
             log.error(t, "Resolve failed. Trying next name server")
-            resolve(name, requestType, tail)
+            resolveWithResolvers(name, requestType, tail)
```
The error when `tail` is `Nil` seems weird but that's not related to this PR, added #26097

raboof commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala:
```diff
@@ -86,63 +96,66 @@ private[io] final class AsyncDnsResolver(
+      // practice, Linux appears to implement something slightly different, if the name being searched contains less
+      // than ndots dots, then it should be searched last, rather than first. This means if the heuristic wrongly
+      // identifies a domain as being relative to the search domains, it will still be looked up if it doesn't resolve
+      // at any of the search domains, albeit with the latency of having to have done all the searches first.
```
Thanks for that extensive comment!

raboof reviewed (commented):
Cool! One question about negative caching, other than that LGTM!

patriknw commented:
@jroper adding mima filters for the internal changes is fine. Please rebase and then I think this is ready to be merged.

patriknw commented:
@raboof do you see anything missing here, otherwise take a final look and approve so we can merge this

raboof commented on akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsCache.scala:
```diff
@@ -32,15 +31,16 @@ import scala.annotation.tailrec
+      case r: ARecord    ⇒ r.ip
+      case r: AAAARecord ⇒ r.ip
     }
+    if (all.isEmpty) None
```
Looking closer, we always check that `answer.rrs.nonEmpty` before putting things in the cache, so we will never put an empty answer in the cache, so returning `None` when there are no records in the cache seems valid.

jroper commented:
Rebased and added problem filters for mima.

jroper commented:
I just pushed a fix to fix the 2.11 build (some diverging implicit expansion on `Ordering[FiniteDuration]` error that I couldn't be bothered really understanding, fixed with a type annotation to make require an `Ordering[Duration]` instead, which seems to work).
